### ISSUE TITLE
Tf.reshape layer readNetFromTensorflow

### ISSUE THREAD

**Issue by xxxcucus** [opencv/opencv#10065]
##### System information (version)

- OpenCV => :master (3.3.1):
- Operating System / Platform => :Linux 64 Bit:
- Compiler => :gcc 5.4:

##### Detailed description
I have the tensor flow model from the following stack overflow question 
https://stackoverflow.com/questions/47072498/evaluation-of-tensorflow-model-with-opencv-fails

The main problem is that while in tensorflow the model works perfectly, with opencv it delivers very low recognition rate (it is about object recognition with a LeNet like network). 

To debug I running the network with a single image both in tensor flow and with opencv. The results are compatible (up to a transposition because 4-tensors in opencv are represented as NCHW and in tensorflow as NHWC) up to the reshape layer. After the reshape layer the order of the data in the tensors is different and I believe this is not taken into account in the next layer when read by opencv.

Here is what tensorflow sees:
[[0.57640886 -0.7320441 -0.56153286 -0.50446403 -0.99916887 -0.25644568 0.37037042 0.35651243 0.98727149 -0.26801953 -0.96697044 0.99657458 0.60084075 -0.947941 -0.9593026 0.80594021 -0.11472013 0.98635733 -0.9991352 -0.64472067 -0.831307 0.99181873 0.9032678 0.23578542 0.93737328 0.7603994 -0.96782762 0.97814667 0.35413632 -0.99387139 -0.98025244 -0.068817958 -0.95615709 -0.79203743 -0.55445671 -0.46434277 -0.53127748 0.42631662 -0.17381959 0.97791088 -0.98697555 -0.2370261 0.70623678 -0.38022324 -0.86536592 -0.56993884 0.84862041 -0.024968814 -0.94662851 -0.92850912 0.6744343 -0.96219629 -0.98179585 -0.013739867 -0.88668972 -0.93075424 -0.99574864 0.41274464 0.27685234 0.1048695 0.80113029 -0.76697719 -0.62179863 -0.54737896 0.88319963 -0.98240131 0.099828944 -0.95020252 -0.99664187 -0.58102214 -0.746525 -0.863225 -0.44986826 0.35669243 -0.92557031 0.92349535 -0.011764921 -0.94197941 -0.95125091 -0.12253148 0.7267735 -0.97859854 -0.88123661 -0.86582363 -0.99845392 0.26621383 -0.25937128 0.16890109 0.8958118 0.46668831 -0.96753418 0.94015944 0.61881793 -0.75414896 -0.93712968 0.67226952 0.93956071 -0.719903 -0.99686307 0.97880363 -0.94701731 0.95968682 0.79836708 0.84152049 0.92731273 0.99217522 -0.89917314 0.31805363 -0.64322454 0.14474566 0.68124747 0.920639 -0.99829763 0.97633177 -0.45174003 0.4868395 0.39509809 0.79704583 0.68296671 -0.79619217 0.62702411 -0.28905591 0.40904808 0.18269578 -0.52435189 -0.77298862 0.95283443 0.77351087 -0.99723041 -0.44964573 0.75458735 -0.84349322 -0.2787348 -0.27797389 -0.98497123 -0.98784333 -0.99876761 0.70629287 0.9589721 0.0011268551 0.59188169 0.065445378 -0.67617232 0.4542411 0.55162752 -0.99224347 0.89628124 -0.90419471 -0.98262393 -0.80243528 -0.97798115 -0.95261228 -0.095537014 0.58084488 -0.42139474 0.43955472 0.13265587 -0.68483931 -0.797325 -0.67334753 0.95912015 -0.73636639 -0.74932885 -0.37403235 -0.99938738 -0.718319 -0.017991962 -0.49010068 0.96980625 0.11315727 -0.96887767 0.99068862 0.3484596 -0.20102145 -0.90014029 0.90703917 0.79230243 0.87842727 -0.99625039 0.99108583 -0.99835312 0.93228185 0.98630559 -0.25911176 0.96960407 0.41488671 -0.99611133 0.96233374 0.73551613 -0.57646978 -0.038171098 0.95043057 -0.99135637 0.40177411 -0.94056273 0.95736593 0.21591909 0.39988357 -0.060853008 0.95882964 -0.1753172 -0.033910573 0.9162333 -0.83249617 -0.99823779 0.72816962 0.99024183 0.012636927 -0.99933159 -0.74018395 0.39596218 -0.56585878 0.77357829 0.38421142 -0.43701345 0.69280165 -0.99971414 -0.85308927 0.77477884 -0.95736516 0.41985902 0.83000326 0.98792142 0.59115183 0.96500641 -0.98292291 0.72910708 -0.59624732 -0.88770288 -0.85520512 -0.79491454 -0.8095634 -0.80206 -0.53263462 -0.93519306 -0.87734377 0.89052463 0.8575471 -0.66406125 -0.87333661 0.45338351 -0.98892277 -0.81889677 -0.7277481 -0.99963719 -0.663199 -0.79925984 -0.97793818 0.74306667 -0.46677661 -0.94928163 0.95268244 0.517398 -0.82154727 -0.92907196 0.20604146 0.742182 -0.94466656 -0.99726743 -0.73200506 -0.99408734 0.86578935 0.53643763 -0.95009625 0.8865959 0.75020939 -0.98732084 0.94216293 0.64727283 -0.99595457 -0.97794354 -0.51353347 0.10758308 -0.655381 -0.97806311 0.037053347 0.82372892 0.98531437 0.92923534 -0.78015614 0.27903911 -0.86078268 0.58897048 -0.5440619 -0.70417482 0.55102015 0.94441944 -0.98200387 -0.96421742 -0.82811987 -0.7621935 -0.028579134 -0.3405354 0.92280525 -0.92901427 -0.16299394 -0.99069488 -0.93840826 -0.15393959 -0.90333724 -0.93428111 0.82822591 0.9926098 -0.99556637 0.73334134 -0.99418527 0.79746032 -0.60684174 -0.93663031 -0.87391633 -0.78419179 -0.517411 -0.78014 0.56145036 -0.98505181 0.97369695 0.95222187 0.957387 0.979379 0.76398283 0.5358116 -0.94371355 -0.89016688 -0.68990403 -0.99937391 0.85284436 -0.97667336 -0.9064464 0.4671514 0.7751866 -0.68927008 0.72050577 0.42299682 -0.9154532 -0.853466 0.4217563 -0.23058192 -0.90318829 -0.996565 -0.77907187 -0.99921858 -0.15937106 -0.93884164 -0.93047738 0.5233016 0.96962792 -0.70453644 0.14469752 -0.63045436 -0.99173361 -0.92830372 0.75817424 -0.61011165 -0.962129 -0.99620408 -0.8788079 -0.9987641 0.13976505 -0.33011863 -0.57233638 0.28096408 0.7246145 -0.41102031 0.14257059 -0.87218511 -0.992732 -0.50290006 -0.67671168 -0.96062219 -0.96131366 -0.95361078 -0.84331673 -0.98659253 -0.0062631438 -0.97765452 -0.36527866 -0.24954209 0.89231664 0.16818769 -0.25742972 -0.98911148 -0.96149635 -0.25759411 -0.65694958 -0.38451055 -0.61090231 -0.731272 -0.89363807 -0.9980306 0.33547831 -0.94922191 -0.91892743 0.50608528 0.75549388 -0.79815388 0.97932446 -0.17126872 -0.98604345 -0.3768613 0.8923431]]

Here is what opencv sees:
0,576409 -0,114721 -0,956157 -0,946629 0,883200 0,726774 0,939560 -0,998298 -0,997230 0,551627 0,959120 0,792303 -0,991356 -0,999332 0,965006 0,453383 0,742182 0,107585 -0,964218 0,733341 0,535812 -0,230583 -0,610110 -0,960622 -0,384511 -0,732044 0,986357 -0,792038 -0,928509 -0,982401 -0,978599 -0,719903 0,976332 -0,449645 -0,992243 -0,736366 0,878427 0,401775 -0,740183 -0,982923 -0,988923 -0,944666 -0,655380 -0,828120 -0,994185 -0,943714 -0,903188 -0,962129 -0,961314 -0,610902 -0,561533 -0,999135 -0,554457 0,674435 0,099829 -0,881237 -0,996863 -0,451742 0,754587 0,896281 -0,749329 -0,996251 -0,940563 0,395961 0,729107 -0,818897 -0,997268 -0,978063 -0,762195 0,797460 -0,890167 -0,996565 -0,996204 -0,953611 -0,731272 -0,504464 -0,644721 -0,464343 -0,962196 -0,950202 -0,865824 0,978804 0,486837 -0,843493 -0,904195 -0,374032 0,991086 0,957366 -0,565859 -0,596247 -0,727748 -0,732006 0,037050 -0,028577 -0,606841 -0,689904 -0,779072 -0,878807 -0,843317 -0,893638 -0,999169 -0,831307 -0,531278 -0,981796 -0,996642 -0,998454 -0,947017 0,395099 -0,278732 -0,982624 -0,999388 -0,998353 0,215916 0,773576 -0,887703 -0,999637 -0,994087 0,823730 -0,340538 -0,936630 -0,999374 -0,999219 -0,998764 -0,986592 -0,998031 -0,256445 0,991819 0,426317 -0,013740 -0,581022 0,266214 0,959687 0,797045 -0,277973 -0,802435 -0,718319 0,932282 0,399883 0,384215 -0,855205 -0,663199 0,865789 0,985314 0,922806 -0,873916 0,852844 -0,159372 0,139765 -0,006264 0,335479 0,370370 0,903268 -0,173819 -0,886690 -0,746525 -0,259372 0,798366 0,682965 -0,984971 -0,977981 -0,017992 0,986306 -0,060851 -0,437012 -0,794914 -0,799260 0,536437 0,929235 -0,929014 -0,784192 -0,976673 -0,938842 -0,330118 -0,977654 -0,949222 0,356512 0,235785 0,977911 -0,930754 -0,863225 0,168901 0,841519 -0,796193 -0,987843 -0,952612 -0,490101 -0,259110 0,958830 0,692801 -0,809564 -0,977938 -0,950096 -0,780157 -0,162994 -0,517409 -0,906446 -0,930477 -0,572336 -0,365279 -0,918928 0,987271 0,937373 -0,986975 -0,995749 -0,449868 0,895812 0,927313 0,627024 -0,998768 -0,095537 0,969806 0,969604 -0,175318 -0,999714 -0,802060 0,743067 0,886596 0,279041 -0,990695 -0,780140 0,467152 0,523301 0,280964 -0,249542 0,506086 -0,268019 0,760399 -0,237025 0,412745 0,356693 0,466688 0,992175 -0,289059 0,706293 0,580845 0,113158 0,414887 -0,033910 -0,853089 -0,532635 -0,466777 0,750210 -0,860783 -0,938408 0,561449 0,775187 0,969628 0,724615 0,892317 0,755494 -0,966970 -0,967827 0,706237 0,276852 -0,925570 -0,967534 -0,899173 0,409044 0,958972 -0,421393 -0,968878 -0,996111 0,916232 0,774779 -0,935193 -0,949282 -0,987321 0,588972 -0,153940 -0,985052 -0,689270 -0,704536 -0,411021 0,168188 -0,798154 0,996575 0,978147 -0,380224 0,104870 0,923495 0,940160 0,318051 0,182695 0,001127 0,439554 0,990689 0,962333 -0,832495 -0,957365 -0,877345 0,952683 0,942163 -0,544063 -0,903337 0,973697 0,720506 0,144697 0,142570 -0,257430 0,979325 0,600841 0,354136 -0,865366 0,801130 -0,011765 0,618818 -0,643227 -0,524352 0,591882 0,132655 0,348459 0,735515 -0,998238 0,419860 0,890525 0,517398 0,647274 -0,704174 -0,934281 0,952222 0,422996 -0,630454 -0,872185 -0,989111 -0,171268 -0,947941 -0,993871 -0,569938 -0,766978 -0,941980 -0,754150 0,144743 -0,772990 0,065445 -0,684840 -0,201021 -0,576468 0,728169 0,830003 0,857547 -0,821547 -0,995955 0,551022 0,828226 0,957387 -0,915453 -0,991734 -0,992732 -0,961496 -0,986043 -0,959303 -0,980252 0,848620 -0,621799 -0,951251 -0,937130 0,681248 0,952834 -0,676173 -0,797325 -0,900140 -0,038170 0,990242 0,987921 -0,664061 -0,929072 -0,977943 0,944419 0,992610 0,979379 -0,853466 -0,928304 -0,502899 -0,257595 -0,376862 0,805940 -0,068817 -0,024968 -0,547379 -0,122532 0,672270 0,920639 0,773510 0,454242 -0,673350 0,907039 0,950430 0,012636 0,591153 -0,873337 0,206042 -0,513532 -0,982004 -0,995566 0,763982 0,421758 0,758175 -0,676712 -0,656949 0,892343 

[ArchiveOpenCVBug.zip](https://github.com/opencv/opencv/files/1460619/ArchiveOpenCVBug.zip)
##### Steps to reproduce
Archive file contains network (without the data loading part) in tensorflow, .pb file loaded in opencv, function used to test the reshape layer in .cpp file.

**Comment by dkurt** [comment]
@xxxcucus, Please try to look at this example of how to manage data layout in TensorFlow:

1. Generate the data
```python
import tensorflow as tf
import numpy as np

# 3x4x2 (height x width  x channels)
inp = tf.placeholder(np.float32, [1, 3, 4, 2], 'input')

# 3x4x5 (height x width  x channels)
conv = tf.layers.conv2d(inp, filters=5, kernel_size=[3, 3], padding='SAME')

# 5x3x4 (channels x height x width)
conv_nchw = tf.transpose(conv, [0, 3, 1, 2])

# 3*4*5
reshaped = tf.reshape(conv_nchw, [3*4*5])

inpData = np.random.standard_normal(inp.shape).astype(np.float32)
with tf.Session() as s:
    s.run(tf.global_variables_initializer())

    saver = tf.train.Saver()
    saver.save(s, 'model.ckpt')
    tf.train.write_graph(s.graph.as_graph_def(), "", "graph.pb")

    [conv_out, conv_nchw_out, reshaped_out] = s.run([conv, conv_nchw, reshaped],
                                                    feed_dict={inp: inpData})

    np.save('inp.npy', inpData)
    np.save('conv_out.npy', conv_out)
    np.save('conv_nchw_out.npy', conv_nchw_out)
    np.save('reshaped_out.npy', reshaped_out)
```

2. Freeze the graph
```bash
python ~/tensorflow/tensorflow/python/tools/freeze_graph.py \
    --input_graph graph.pb \
    --input_checkpoint model.ckpt \
    --output_graph frozen_graph.pb \
    --output_node_names "Reshape"
```

3. Load in OpenCV
```python
import cv2 as cv
import numpy as np

net = cv.dnn.readNetFromTensorflow('frozen_graph.pb')

inp = np.load('inp.npy')  # (height x width x channels)
inp = np.transpose(inp, [0, 3, 1, 2])  # (channels x height x width)

conv = np.load('conv_out.npy')  # (height x width x channels)
conv = np.transpose(conv, [0, 3, 1, 2])  # (channels x height x width)

conv_nchw = np.load('conv_nchw_out.npy')  # (channels x height x width)
reshaped = np.load('reshaped_out.npy')

net.setInput(inp)

print np.max(np.abs(conv - net.forward('conv2d/convolution')))
print np.max(np.abs(conv_nchw - net.forward('transpose')))
print np.max(np.abs(reshaped - net.forward('Reshape')))
```

**Comment by GPhilo** [comment]
@dkurt about the example (which I can't run just yet because I don't have the right setup on this machine): The TF graph expects input in NHWC format, while `inp` in the opencv part has format NCHW. Is opencv taking care of internally converting everything to NCHW? If so, how is the `transpose(NHWC to NCHW)` in the graph handled?

**Comment by xxxcucus** [comment]
@GPhilo see part of the answer here http://answers.opencv.org/question/177485/testing-tensorflow-model-with-opencv-dnn/

**Comment by xxxcucus** [comment]
@dkurt Thank you for the prompt answer. I added a tf.transpose before the tf.reshape as you suggest in your example and both models (opencv and tensorflow) infer identically now.

**Comment by GPhilo** [comment]
@xxxcucus what data format do you use in Tensorflow for your training?

**Comment by xxxcucus** [comment]
@GPhilo images are casted to tf.float32

**Comment by xxxcucus** [comment]
@dkurt Adding the tf.transpose() in my model has definitely an impact of how my model performs on the training and test data.

**Comment by xxxcucus** [comment]
@GPhilo I am using NHWC for training. I will try in the next days to test with NCHW to optimize for GPU.

**Comment by dkurt** [comment]
@xxxcucus, Could you please test the model again using changes from https://github.com/opencv/opencv/pull/10364?

**Comment by xxxcucus** [comment]
Yes. I will try to test tomorrow or on Friday.

**Comment by xxxcucus** [comment]
@dkurt I have built your pull request but have not had the time to recreate my model. I will continue next week as Christmas duties are overwhelming me right now.

**Comment by xxxcucus** [comment]
@dkurt First indirect tests without tf.transpose are positive. I am testing more times to be sure.

**Comment by xxxcucus** [comment]
Tested also directly that opencv delivers the same results after softmax as tensorflow model.
